### DISCUSSION TITLE
Implement Leap Key Navigation

### DISCUSSION THREAD

**Discussion by peter-wasilko** [zed-industries/zed#25470] [Feature Requests]
### Check for existing issues

- [X] Completed

### Describe the feature

Jef Raskin's design for the Canon Cat employed a quasi-mode for rapid navigation as illustrated in this [YouTube Video](https://www.youtube.com/watch?v=o_TlE_U_X3c&t=212s).

[This feature](https://github.com/helloSystem/hello/issues/55) has also been requested in the helloSystem project repro.

Also, BitSavers has a copy of the Cat's [Technical Documentation](http://www.bitsavers.org/pdf/informationAppliance/cat/Technical_Documentation_for_the_Canon_Cat_Editor_Sep88.pdf) that may be of interest on this topic among others.

Finally, there was an attempted [reimplementation](https://en.wikipedia.org/wiki/Archy_(software)) in Python by Aza Raskin, Jef's son.

**Comment by clo4** [comment]
This is an interesting idea. I don't see how it's different from using cmd-f though (or `/`/`?` in vim mode) -- could you explain further what makes this different from just searching?

Seems similar to zed-industries/zed#4930 and zed-industries/zed#5289 (it's a different execution of the same concept)

**Reply by 0xDAEF0F** [reply]
using `/` or `?` is much much slower and inefficient.

**Comment by clord** [comment]
Plus one for this. 

I prefer [leap](https://github.com/ggandor/leap.nvim) over [hop](https://github.com/hadronized/hop.nvim) and [easymotion](https://github.com/easymotion/vim-easymotion) personally, as it is very natural; using the word prefix itself instead of random characters. With it, you trigger the command somehow and then press the two letters of the word you want the cursor to go, and all locations that match on screen get an indication. since you're looking at where you want to go, you press the key that marks the location, and your cursor jumps. Since only places with the two characters appear, the visual clutter is minimal.

Please see the readme at https://github.com/ggandor/leap.nvim

Other editors have this idea too, and bring new concepts to it. For instance, helix recently added `goto_word` mode. 

This category of plugin speaks to the need for quickly moving the cursor visually without touching the mouse or changing modes.

**Comment by mmosk** [comment]
Currently, [leap.nvim](https://github.com/ggandor/leap.nvim) and [oil.nvim](https://github.com/stevearc/oil.nvim) are the two things I miss the most, in that order.

I like `/` for being stateful (cycling through matches) and global, but when you just need to jump to a certain place in a viewport (which I believe to be the most common motion), it seems like nothing does it better than leap.nvim at the moment.

**Comment by willtalmadge** [comment]
> This is an interesting idea. I don't see how it's different from using cmd-f though (or `/`/`?` in vim mode) -- could you explain further what makes this different from just searching?
> 
> Seems similar to #4930 and #5289 (it's a different execution of the same concept)

You can think of it like a poor-man's eye tracker because it serves the same functional purpose. If you are looking at a location in a vim buffer, you can get the cursor to that location by pressing no more than 2-3 keys (not including the start key). So you look at something,  press 's', start typing what you are looking at, if there's ambiguity it'll temporarily put a "flag" character in the buffer you should type to disambiguate which site you are looking at. It feels like your editor is reading your mind. Completely replaces the need for touching the trackpad which is helpful for me because trackpad clicking gives me carpal tunnel.

**Comment by ninjarogue** [comment]
Here's a quick demo:

https://github.com/user-attachments/assets/ca821e97-3460-4923-925f-08cb6f5d341a

I'm pressing 's' to activate leap.

**Comment by djmango** [comment]
Is this possible to add via an extension? Would prefer core ideally but I'd be happy to look into writing a plugin

**Comment by clord** [comment]
This is one of those things that Zed could really shine at, since it is not limited to the character grid like a terminal editor. Target annotations could avoid replacing characters at the target. 

Another noteworthy thing for this feature is that it works cross-tab. In vim, I set up `S` to jump to another split, and then plopping the cursor right where it's needed.

**Comment by trkoch** [comment]
This feature request might seem like a stretch and could appear arbitrary to implement since it's just one of many Vim plugins. However, consider that Zed is uniquely positioned to implement this seamlessly (implementation in vscode-neovim is decent, but limited). I agree and concur `leap` greatly enhances keyboard navigation, aligning with Zed’s focus on minimizing mouse use.

**Comment by PlexSheep** [comment]
Coming from neovim and giving zed a real chance, this is something I really miss.

For one, in the actual editing buffer, but also between different buffers/panels/whatever

It's just a very efficient way to move without making use of a mouse.

**Comment by Oyelowo** [comment]
Just to add to these, I think something like [flit.nvim](https://github.com/ggandor/flit.nvim) should be considered in addition to [leap.nvim](https://github.com/ggandor/leap.nvim) if this ever gets implemented.   

[flit.nvim](https://github.com/ggandor/flit.nvim) builds on  [leap.nvim](https://github.com/ggandor/leap.nvim) but I use its(`flit's`) code navigation `>95%` of the times and use [leap.nvim](https://github.com/ggandor/leap.nvim) mainly when i need to navigate across panes or blocks with a lot of character repetitions.

I have just tried zed yet again and I love the smoothness and the feel. That said, [oil.nvim](https://github.com/stevearc/oil.nvim), [flit.nvim](https://github.com/ggandor/flit.nvim) and [leap.nvim](https://github.com/ggandor/leap.nvim) are three plugins I have come to rely on heavily and I think people might appreciate their approaches.

**Comment by 0xDAEF0F** [comment]
Helix editor has this feature integrated natively via `goto word`. Here is where it is mentioned: https://docs.helix-editor.com/keymap.html?search=goto%20word

And yes, it is a killer feature. I'm currently using [flash](https://github.com/folke/flash.nvim) over leap which I think it's a bit better but yeah. same thing.

Really hope this get's implemented somehow.

**Comment by Error6929** [comment]
So, what's the status on this?

**Reply by vicolby** [reply]
Give me a week

**Reply by liamwh** [reply]
How'd it go? Would love it!

**Comment by i-void** [comment]
Is there any update on this? It would be a great enhancement for vim/nvim users.

**Comment by silenceofthewans** [comment]
This would be a solved problem if they allowed embedding neovim just like vscode-neovim extension but for some reasons mentioned here https://zed.dev/blog/zed-decoded-vim they didn't and it sounds a lot like "not invented here" they could've just exposed some api so someone could make a zed-neovim extension while keeping intact their "dna" of crdt based collaboration but the current way of basically re-implementing vim/neovim inside zed would always leave something out and would never be complete.

**Reply by kmicklas** [reply]
There is no inherent relation between this feature and vim. I do not use vim keybindings and I would like leap motion functionality. Having to use an embedded vim to get that would be crazy.

**Comment by MadeInShineA** [comment]
Any update on this feature ? I feel like that's on of the only thing holding me back...

**Reply by laurhinch** [reply]
As far as I can tell, multiple people have branches containing the work for this feature (or a similar feature), but they are all held up behind work on a consistent glyph overlay system and design style, which has yet to be developed.

**Comment by JosephTLyons** [comment]
Just to note: Helix amp jump navigation has shipped in [Zed Stable 1.1.5](https://zed.dev/releases/stable).
This shipped as a Helix-mode feature bound to `g w`, rather than as an editor-wide Leap-style command.
Thanks to [@GoldStrikeArch](https://github.com/GoldStrikeArch) for the implementation.

Ill leave this discussion open for now since this isn't the exact ask, and this implementation currently only works in Vim or Helix mode, and isn't available outside of that, but users of those modes might be interested to try it out.

**Reply by Error6929** [reply]
👍🏻
